federationtypes.go

```diff
@@ -347,9 +351,18 @@ func (r *RespSendJoin) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// ToRespState return new RespState with the same data from the given RespSendJoin
+func (r RespSendJoin) ToRespState() RespState {
+	return RespState{
+		StateEvents: r.StateEvents,
+		AuthEvents:  r.AuthEvents,
+	}
+}
+
 type respSendJoinFields struct {
```

Comment:
Since this struct is just an adaptor between the original `RespState` and `RespSendJoin`, it can now be removed. Just move these field tags (`json:"..."`) over to the `RespSendJoin` struct and we should be good to go!

Reply:
After trying to get rid of `respSendJoinFields `  i can't find a way to properly make `RespSendJoin` implement `json.MarshalJSON` and `json.UnmarshalJSON` interfaces without a type that is indirectly like `RespSendJoin`. 

 Currently the way `RespSendJoin` implement `json.MarshalJSON` and `json.UnmarshalJSON` interfaces, i think `respSendJoinFields` is still needed. 

I also found the same technique being used on `RespInvite`, i think this is because we have to deal with an array that hold data of differents types. 

However i'am still trying to find a work around.

Reply:
Oh, so there's going to be a recursive call if we remove this struct, is this what you meant? If that's the case, then I think we can leave the struct there for now. A comment on that can be added to the struct still, for clarity.

Reply:
That's exactly what I meant